### ISSUE TITLE
java.lang.NoClassDefFoundError: com/sun/jna/platform/win32/Shell32

### ISSUE THREAD

**Issue by erayerdin** [harawata/appdirs#2]
I am using `appdirs` in a JavaFX application. appdirs throws exception in Windows 7 as below:

```
Exception in thread "JavaFX Application Thread" Exception in thread "main" java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at sun.launcher.LauncherHelper$FXHelper.main(Unknown Source)
Caused by: java.lang.NullPointerException
        at com.sun.javafx.application.LauncherImpl.launchApplicationWithArgs(LauncherImpl.java:383)
        at com.sun.javafx.application.LauncherImpl.launchApplication(LauncherImpl.java:328)
        ... 5 more
java.lang.NoClassDefFoundError: com/sun/jna/platform/win32/Shell32
        at net.harawata.appdirs.impl.ShellFolderResolver.resolveFolder(ShellFolderResolver.java:38)
        at net.harawata.appdirs.impl.WindowsAppDirs.getLocalAppData(WindowsAppDirs.java:71)
        at net.harawata.appdirs.impl.WindowsAppDirs.getUserDataDir(WindowsAppDirs.java:33)
        at net.harawata.appdirs.AppDirs.getUserDataDir(AppDirs.java:28)
        at com.erayerdin.corpustk.App.<clinit>(App.java:19)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Unknown Source)
        at com.sun.javafx.application.LauncherImpl.lambda$launchApplicationWithArgs$155(LauncherImpl.java:352)
        at com.sun.javafx.application.PlatformImpl.lambda$runAndWait$174(PlatformImpl.java:326)
        at com.sun.javafx.application.PlatformImpl.lambda$null$172(PlatformImpl.java:295)
        at java.security.AccessController.doPrivileged(Native Method)
        at com.sun.javafx.application.PlatformImpl.lambda$runLater$173(PlatformI
mpl.java:294)
        at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:95)
        at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
        at com.sun.glass.ui.win.WinApplication.lambda$null$147(WinApplication.java:177)
        at java.lang.Thread.run(Unknown Source)
Caused by: java.lang.ClassNotFoundException: com.sun.jna.platform.win32.Shell32
        at java.net.URLClassLoader.findClass(Unknown Source)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        at sun.misc.Launcher$AppClassLoader.loadClass(Unknown Source)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        ... 16 more
```

It works in Linux, though. Adding `com.sun.jna` as a Maven dependency does not work as well.

***

# Environment

 - Microsoft Windows [Version 6.1.7601] (stands for Windows 7)
 - Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
 - [Appdirs 1.0.0](https://mvnrepository.com/artifact/net.harawata/appdirs/1.0.0)

**Comment by harawata** [comment]
Hi @erayerdin ,

`com.sun.jna` seems to be very old.
Could you try adding the following dependency if you haven't?

```xml
<dependency>
    <groupId>net.java.dev.jna</groupId>
    <artifactId>jna</artifactId>
    <version>4.5.0</version>
</dependency>
```

**Comment by erayerdin** [comment]
Yeah, I also saw that. I hesitated to try it since the library *specifically* calls `com.sun.jna`, not `net.java.dev.jna`.

However, I will give it a try and post the results here.

**Comment by erayerdin** [comment]
Update: I was right. The same error occurs. My debug log below:

```
Exception in thread "JavaFX Application Thread" Exception in thread "main" java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at sun.launcher.LauncherHelper$FXHelper.main(Unknown Source)
Caused by: java.lang.NullPointerException
        at com.sun.javafx.application.LauncherImpl.launchApplicationWithArgs(LauncherImpl.java:383)
        at com.sun.javafx.application.LauncherImpl.launchApplication(LauncherImpl.java:328)
        ... 5 more
java.lang.NoClassDefFoundError: com/sun/jna/platform/win32/Shell32
        at net.harawata.appdirs.impl.ShellFolderResolver.resolveFolder(ShellFolderResolver.java:38)
        at net.harawata.appdirs.impl.WindowsAppDirs.getLocalAppData(WindowsAppDirs.java:71)
        at net.harawata.appdirs.impl.WindowsAppDirs.getUserDataDir(WindowsAppDirs.java:33)
        at net.harawata.appdirs.AppDirs.getUserDataDir(AppDirs.java:28)
        at com.erayerdin.corpustk.App.<clinit>(App.java:19)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Unknown Source)
        at com.sun.javafx.application.LauncherImpl.lambda$launchApplicationWithArgs$155(LauncherImpl.java:352)
        at com.sun.javafx.application.PlatformImpl.lambda$runAndWait$174(PlatformImpl.java:326)
        at com.sun.javafx.application.PlatformImpl.lambda$null$172(PlatformImpl.
java:295)
        at java.security.AccessController.doPrivileged(Native Method)
        at com.sun.javafx.application.PlatformImpl.lambda$runLater$173(PlatformI
mpl.java:294)
        at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatch
er.java:95)
        at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
        at com.sun.glass.ui.win.WinApplication.lambda$null$147(WinApplication.ja
va:177)
        at java.lang.Thread.run(Unknown Source)
Caused by: java.lang.ClassNotFoundException: com.sun.jna.platform.win32.Shell32
        at java.net.URLClassLoader.findClass(Unknown Source)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        at sun.misc.Launcher$AppClassLoader.loadClass(Unknown Source)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        ... 16 more
```

I will, still again, check my code to see if there is a mistake I've been doing (oddly, it works on Ubuntu even if there is a mistake).

**Comment by harawata** [comment]
That error means your application could not find `jna-4.5.0.jar` in the classpath.
It occurs when, for example, the application is built with some incorrect build setting.

First, please make sure that `jna-4.5.0.jar` is included in your final JavaFX application.
If your application is a single .jar file, unpack it with `jar -xvf` or change the extension to `.zip` and unzip it.

If `jna-4.5.0.jar` is included and you still get the error, there might be something wrong with the application's MANIFEST.MF.
See if `jna-4.5.0.jar` is listed in `JavaFX-Application-Class`.

If the above does not help, please let me know how you build your JavaFX application (i.e. using IDE? mvn command?, etc.).

p.s.
JNA is used only on Windows, so it is not surprising that it works on Ubuntu.

**Comment by erayerdin** [comment]
Project already uses Maven, so it configures the classpath well. The project is [here](https://github.com/erayerdin/corpustk).

I managed to define application data folders manually right now, so my problem is kind of solved. If I manage to find any further resources (like debug logs, code lines et cetera), I will give further information.

**Comment by harawata** [comment]
You seem to be using IntelljiJ as your IDE.
I'm not familiar with it at all, but open Project Settings -> Modules -> Dependencies and see if  'Export' is checked for the JNA library.

**Comment by zak-pawel** [comment]
@erayerdin 
You need not only dependency to **jna**, but also to **jna-platform**.

Probably this should solve your problem:
```xml
<dependency>
    <groupId>net.java.dev.jna</groupId>
    <artifactId>jna</artifactId>
    <version>4.5.0</version>
</dependency>
<dependency>
    <groupId>net.java.dev.jna</groupId>
    <artifactId>jna-platform</artifactId>
    <version>4.5.0</version>
</dependency>
```

**Comment by erayerdin** [comment]
@zak-pawel and @harawata 
First, thanks for your quick response and focusing on this problem. However, due to my time limitation, I stopped utilizing this library in my project and patched manually by checking `user.home` stuff and so.

Since I dropped the library, I cannot give you further detail. Yet again, Maven is also a dependency management tool apart from building tool, so the configuration of this project must provide other dependencies and then be updated in Maven repository. Please, don't get me wrong. This is not an advice, this is just a suggestion from a random stranger on the internet.

This is the only library in Java and Maven repository utilizing crossplatform application directory management. I hope someone else will have the problem and can give you further details right after implementing your explanations.

Thanks again and I am sorry I could not continue using this library and give you enough feedback since I don't have enough time.

**Comment by zak-pawel** [comment]
@erayerdin 
I'm not co-author of this library. I'm just a random stranger on the internet who had the same problem as you and found a workaround for this issue.

I totally agree that configuration of this project should provide other dependencies and I hope that @harawata fixes this dependency issue.

**Comment by harawata** [comment]
@erayerdin ,

No worries. I appreciate you took the time to report this!

@zak-pawel is right (thanks for the help!). I checked the dependency and it is `jna-platform` that is required by appdirs (`jna` is also required, but it will be automatically added because `jna-platform` depends on it).

Anyway, to make things simpler, I have modified the scope of `jna-platform` dependency to `compile` and released 1.0.1.
